urld2fb1d2b05_thnaeff_Util_tgz-pJ8-ch_thn_util_tests_TreeTestJ8/src/ch/thn/util/file/PollingWatchService.java

```diff
@@ -0,0 +1,456 @@
+	private long pollTimeout = 0;
+
+	private ConcurrentHashMap<Path, PollingWatchKey> registeredPaths = null;
+
+	/**
+	 * A directory and all the files and their last modified timestamps<br />
+	 * &lt;Directory, &lt;File path, last modified timestamp&gt;&gt;
+	 */
+	private ConcurrentHashMap<Path, Map<Path, Long>> lastModified = null;
+
+	/**
+	 * A FIFO list of all the watch keys which have events pending
+	 */
+	private LinkedBlockingQueue<PollingWatchKey> keysWithEvents = null;
+
+	private FilenameFilter fileNameFilter = null;
+
+	/**
+	 * 
+	 * 
+	 * @param pollTimeout The timeout to wait between two file checks.
+	 */
+	protected PollingWatchService(long pollTimeout) {
+		this(pollTimeout, null);
+	}
+
+	/**
+	 * 
+	 * 
+	 * @param pollTimeout The timeout to wait between two file checks.
+	 * @param fileNameFilter A filter to only check the filtered files. This can
+	 * be a big performance improvements when dealing with large and/or many directories
+	 */
+	protected PollingWatchService(long pollTimeout, FilenameFilter fileNameFilter) {
+		super(true, true);
+
+		this.pollTimeout = pollTimeout;
+		this.fileNameFilter = fileNameFilter;
+
+		registeredPaths = new ConcurrentHashMap<>();
+		lastModified = new ConcurrentHashMap<>();
+		keysWithEvents = new LinkedBlockingQueue<PollingWatchKey>();
+
+	}
+
+
+
+	@Override
+	public void run() {
+		runStart();
+
+		while (!isStopRequested()) {
+			runReset();
+			runPause(false);
+
+			if (isStopRequested()) {
+				break;
+			}
+
+			if (isPauseRequested() || isResetRequested()) {
+				continue;
+			}
+
+			lookForChanges();
+
+			controlledWait(pollTimeout);
+
+		}
+
+		registeredPaths.clear();
+		lastModified.clear();
+		keysWithEvents.clear();
+
+		for (PollingWatchKey key : registeredPaths.values()) {
+			key.cancel();
+		}
+
+		notifyWaitingTake();
+
+		runEnd();
+	}
+
+
+	/**
+	 * 
+	 * 
+	 */
+	private void lookForChanges() {
+
+		//Set<Path> tempPaths = new HashSet<>();
+
+		for (Path directoryPath : registeredPaths.keySet()) {
+			//If a registered path gets deleted because it does not exist any more
+			//and it is the next path in line, it still gets returned. Check here again
+			//if it it still available.
+			if (!registeredPaths.containsKey(directoryPath)) {
+				continue;
+			}
+
+			File directory = directoryPath.toFile();
+			PollingWatchKey key = registeredPaths.get(directoryPath);
+
+
+			//Does registered directory still exist?
+			if (!directory.exists()) {
+				entryDelete(key, directoryPath);
+				continue;
+			}
+
+			File[] files = null;
+			if (fileNameFilter != null) {
+				files = directory.listFiles(fileNameFilter);
+			} else {
+				files = directory.listFiles();
+			}
+
+			if (!lastModified.containsKey(directoryPath)) {
+				//It is a new path which has just been added and has not been checked yet.
+				//Record all files and their modified timestamp
+
+				Map<Path, Long> filesMap = new ConcurrentHashMap<Path, Long>();
+				lastModified.put(directoryPath, filesMap);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `containsKey()` and `put()` may not be thread-safe at lines: **167 and 172**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread&#39;s `putIfAbsent()`.

**More info**  
 The thread whose value was overwritten still has a reference to that value and uses that reference at line: **178**. `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.